### PR TITLE
Expose raw CDR stream for publish and subscribe

### PR DESCRIPTION
**Pull request by Karsten1987**
Connects to ros2/demos#185

### CODE REVIEW

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp [hunk 1]
```diff
@@ -764,7 +760,14 @@ bool TypeSupport<MembersType>::deserialize(
   assert(payload);
 
   auto buffer = static_cast<eprosima::fastcdr::FastBuffer *>(data);
-  buffer->resize(payload->length);
+  // TODO(karsten1987): This is a bug IMO
+  // see issue here: https://github.com/eProsima/Fast-CDR/issues/15
+  // an empty buffer will always be initialized to a hard set value - atm 200
+  // it will call malloc internally, so we can use malloc here to preinitialize our
+  // own buffer.
+  if (!buffer->reserve(payload->length)) {
```
Comment (sloretz) [inline rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp]: I get a compiler error here

```
In file included from /workspace/ros2-pr-review/src/ros2/rmw_fastrtps/rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport
.hpp:146:0,
                 from /workspace/ros2-pr-review/src/ros2/rmw_fastrtps/rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/MessageType
Support.hpp:24,
                 from /workspace/ros2-pr-review/src/ros2/rmw_fastrtps/rmw_fastrtps_cpp/src/type_support_common.hpp:25,
                 from /workspace/ros2-pr-review/src/ros2/rmw_fastrtps/rmw_fastrtps_cpp/src/client_service_common.cpp:17:
/workspace/ros2-pr-review/src/ros2/rmw_fastrtps/rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp: In member f
unction ‘bool rmw_fastrtps_cpp::TypeSupport<MembersType>::deserialize(eprosima::fastrtps::rtps::SerializedPayload_t*, void*
)’:
/workspace/ros2-pr-review/src/ros2/rmw_fastrtps/rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp:768:16: erro
r: ‘class eprosima::fastcdr::FastBuffer’ has no member named ‘reserve’
   if (!buffer->reserve(payload->length)) {
```

**Comment by Karsten1987** [comment]
you need to checkout this PR as well
https://github.com/eProsima/Fast-CDR/pull/16

File: rmw_fastrtps_cpp/src/rmw_publish.cpp [hunk 2]
```diff
@@ -58,4 +58,35 @@ rmw_publish(const rmw_publisher_t * publisher, const void * ros_message)
...
+  }
+
+  auto info = static_cast<CustomPublisherInfo *>(publisher->data);
+  assert(info);
+
+  eprosima::fastcdr::FastBuffer buffer(raw_message->buffer, raw_message->buffer_length);
+  eprosima::fastcdr::Cdr ser(
+    buffer, eprosima::fastcdr::Cdr::DEFAULT_ENDIAN, eprosima::fastcdr::Cdr::DDS_CDR);
+  if (!ser.jump(raw_message->buffer_length)) {
+    RMW_SET_ERROR_MSG("cannot correctly set raw buffer");
```
[the remark this answers is not in the document]
Comment (Karsten1987) [reply rmw_fastrtps_cpp/src/rmw_publish.cpp]: done

File: rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/macros.hpp [hunk 3]
```diff
@@ -33,4 +33,12 @@
     } \
   };
 
+// check the input pointer for null, sets the error msg accordingly and
+// return RMW_RET_ERROR if null
+#define RETURN_ERROR_ON_NULL(POINTER) \
```
Comment (sloretz) [inline rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/macros.hpp]: Could use `RCUTILS_CHECK_FOR_NULL_WITH_MSG` [link](https://github.com/ros2/rcutils/blob/9092a2581aba9be2a44f90a646e75f8d129ccdfa/include/rcutils/error_handling.h#L128).

**Pull request merged**